static/js/reactions.js

```diff
@@ -0,0 +1,176 @@
+        return reaction.emoji_name === emoji_name;
+    });
+    return matching_reactions.map(function (reaction) {
+        return reaction.user.id;
+    });
+}
+
+exports.message_reaction_on_click = function (message_id, emoji_name) {
+    // When a message's reaction is clicked,
+    // if the user has reacted to this message with this emoji
+    // the reaction is removed
+    // otherwise, the reaction is added
+    var user_list = get_user_list_for_message_reaction(message_id, emoji_name);
+    var operation = 'remove';
+    if (user_list.indexOf(page_params.user_id) === -1) {
+        // User hasn't reacted with this emoji to this message
+        operation = 'add';
+    }
+    send_reaction_ajax(message_id, emoji_name, operation);
+};
+
+function reaction_popover_reaction_on_click() {
+    // When an emoji is clicked in the popover,
+    // if the user has reacted to this message with this emoji
+    // the reaction is removed
+    // otherwise, the reaction is added
+    var emoji_name = this.title;
+    var message_id = $(this).parent()[0].title;
+    var user_list = get_user_list_for_message_reaction(message_id, emoji_name);
+    var operation = 'add';
+    if (user_list.indexOf(page_params.user_id) !== -1) {
+        // User has reacted with this emoji to this message
+        $(this).removeClass('reacted');
+        operation = 'remove';
+    }
+    send_reaction_ajax(message_id, emoji_name, operation);
+    popovers.hide_all();
+}
+
+function filter_emojis() {
+    var search = $(".reaction-popover-filter");
+    var search_term = search.expectOne().val().trim();
+    var reaction_list = $(".reaction-popover-reaction");
+    if (search_term !== '') {
+        reaction_list.filter(function () {
+            return this.title.indexOf(search_term) === -1;
+        }).css("display", "none");
+        reaction_list.filter(function () {
+            return this.title.indexOf(search_term) !== -1;
+        }).css("display", "block");
+    } else {
+        reaction_list.css("display", "block");
+    }
+}
+
+$(document).on('click', '.reaction-popover-reaction', reaction_popover_reaction_on_click);
+$(document).on('input', '.reaction-popover-filter', filter_emojis);
+
+function full_name(user_id) {
+    if (user_id === page_params.user_id) {
+        return 'You (click to remove)';
+    }
+    return people.get_person_from_user_id(user_id).full_name;
+}
+
+function generate_title(emoji_name, user_ids) {
+    var i = user_ids.indexOf(page_params.user_id);
+    if (i !== -1) {
+        // Move current user's id to start of list
+        user_ids.splice(i);
+        user_ids.unshift(page_params.user_id);
+    }
+    return user_ids.map(full_name).join(', ') + ' reacted with :' + emoji_name + ':';
+}
+
+exports.add_reaction = function (event) {
+    var emoji_name = event.emoji_name;
+    var message_id = event.message_id;
+    event.user.id = event.user.user_id;
+    message_store.get(message_id).reactions.push(event);
```

Comment:
Re: the first comment, messages in message_store have a reactions property that is a list of tuples (emoji_name, user_id). This is what get_old_messages sends from the server, a list of individual reactions (reactions in the database are unique on message_id, emoji_name and user_id). 
Adding a new reaction (emoji_name, user_id) to this list is what we want to do.